messages.py

```diff
@@ -68,120 +66,131 @@ def handle(self, response):
-                user.offenses += 1
+            else:
+                user = User(id=data["user_id"], joins=1, messages=1)
+                session.add(user)
                 session.commit()
-                return self.send_message(
-                    data["user_name"], "Please stop spamming emoticons.",
-                    method="whisper")
-            elif (findall(("http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|"
-                           "[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+"),
-                          parsed) and not
-                    self.config["spam_protection"].get(
-                        "allow_links", False)):
+
+            mod_roles = ("Owner", "Staff", "Founder", "Global Mod", "Mod")
+            if not (data["user_roles"][0] in mod_roles or user.friend):
+                if (len(parsed) > self.config["spam_protection"].get(
+                        "maximum_message_length", 256)):
+                    self.remove_message(data["channel"], data["id"])
+                    user.offenses += 1
+                    session.commit()
+                    return self.send_message(
+                        data["user_name"], "Please stop spamming.",
+                        method="whisper")
+                elif (sum(char.isupper() for char in parsed) >
+                        self.config["spam_protection"].get(
+                            "maximum_message_capitals", 32)):
+                    self.remove_message(data["channel"], data["id"])
+                    user.offenses += 1
+                    session.commit()
+                    return self.send_message(
+                        data["user_name"], "Please stop speaking in all caps.",
+                        method="whisper")
+                elif (sum(chunk["type"] == "emoticon"
+                          for chunk in data["message"]["message"]) >
+                        self.config["spam_protection"].get(
+                        "maximum_message_emotes", 8)):
+                    self.remove_message(data["channel"], data["id"])
+                    user.offenses += 1
+                    session.commit()
+                    return self.send_message(
+                        data["user_name"], "Please stop spamming emoticons.",
+                        method="whisper")
+                elif (findall((r"http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|"
+                              r"[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+"),
+                              parsed) and not
+                        self.config["spam_protection"].get(
+                            "allow_links", False)):
+                    self.remove_message(data["channel"], data["id"])
+                    user.offenses += 1
+                    session.commit()
+                    return self.send_message(
+                        data["user_name"], "Please stop posting links.",
+                        method="whisper")
+
+            if parsed == "/cry":
```

Comment:
We should probably remove this... no longer works with Beam's frontend.

.../cry :sob: